blockchain/blockchain.go

```diff
@@ -0,0 +1,316 @@
+
+	// Use the bufPool and reqPool to reuse bytes.Buffer and http.Request objects
+	buf := bl.bufPool.Get().(*bytes.Buffer)
+	req := bl.reqPool.Get().(*http.Request)
+	defer func() {
+		bl.putBuf(buf)
+		bl.putReq(req)
+	}()
+
+	if err := json.NewEncoder(buf).Encode(p); err != nil {
+		return nil, err
+	}
+	req, err := http.NewRequestWithContext(ctx, method, addr, buf)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+
+	resp, err := bl.ch.Do(req)
+
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+
+	var bufRes bytes.Buffer
+	_, err = io.Copy(&bufRes, resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	b := bufRes.Bytes()
+	switch {
+	case resp.StatusCode != http.StatusOK:
+		return nil, fmt.Errorf("unexpected response: %d %s", resp.StatusCode, string(b))
+	default:
+		return b, nil
+	}
+}
+
+func (bl *FxBlockchain) serve(w http.ResponseWriter, r *http.Request) {
+	from, err := peer.Decode(r.RemoteAddr)
+	if err != nil {
+		log.Debug("cannot parse remote addr as peer ID: %v", err)
+		http.Error(w, "", http.StatusBadRequest)
+		return
+	}
+	action := path.Base(r.URL.Path)
+	if !bl.authorized(from, action) {
+		log.Debug("rejected unauthorized request from %s for action %s", from, action)
+		http.Error(w, "", http.StatusUnauthorized)
+		return
+	}
+	// Define a map of functions with the same signature as handleAction
+	actionMap := map[string]func(peer.ID, http.ResponseWriter, *http.Request){
+		actionSeeded: func(from peer.ID, w http.ResponseWriter, r *http.Request) {
+			bl.handleAction(actionSeeded, from, w, r)
+		},
+		actionAuth: func(from peer.ID, w http.ResponseWriter, r *http.Request) {
+			bl.handleAuthorization(from, w, r)
+		},
+	}
+
+	// Look up the function in the map and call it
+	handleActionFunc, ok := actionMap[action]
+	if !ok {
+		http.Error(w, "", http.StatusNotFound)
+		return
+	}
+	handleActionFunc(from, w, r)
+}
+
+func (bl *FxBlockchain) handleAction(action string, from peer.ID, w http.ResponseWriter, r *http.Request) {
+	log := log.With("action", action, "from", from)
+	req := reflect.New(requestTypes[action]).Interface()
+	res := reflect.New(responseTypes[action]).Interface()
+
+	defer r.Body.Close()
+
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		log.Debug("cannot parse request body: %v", err)
+		http.Error(w, "", http.StatusBadRequest)
+		return
+	}
+
+	//go func() {
+	ctx, cancel := context.WithTimeout(r.Context(), time.Second*30)
+	defer cancel()
+	response, err := bl.callBlockchain(ctx, action, req)
+	if err != nil {
+		log.Error("failed to process seeded request: %v", err)
+		return
+	}
+	w.WriteHeader(http.StatusAccepted)
+
+	err1 := json.Unmarshal(response, &res)
+	if err1 != nil {
+		log.Error("failed to format response: %v", err1)
+	}
+
+	if err := json.NewEncoder(w).Encode(res); err != nil {
+		log.Error("failed to write response: %v", err)
+	}
```

Comment:
I am curious why bother with unmarshal and encode when we don't do anything with `res`?